### PR TITLE
Implement `IntoIterator` for `VariableSizeListArray`

### PR DESCRIPTION
**Pull request by dhruv9vats**
addresses https://github.com/mbrobbel/narrow/issues/181

### CODE REVIEW

File: src/offset.rs [hunk 1]
```diff
@@ -490,6 +490,125 @@ where
...
 
+/// An owned iterator over an offset
+pub struct OffsetIntoIter<T, OffsetItem: OffsetElement, Buffer: BufferType>
+where
+    T: IntoIterator,
+{
+    /// The underlying array data
+    data: <T as IntoIterator>::IntoIter,
+    /// The offsets buffer
+    offsets: <Buffer as BufferType>::Buffer<OffsetItem>,
```
Comment (mbrobbel) [inline src/offset.rs:501]: Can't we also use the `Buffer`'s `IntoIterator` instead of relying on indexing? You can wrap it in a [`Peekable`](https://doc.rust-lang.org/stable/std/iter/struct.Peekable.html). Then we don't need to keep `index`, but we just drain the `offsets` iterator.

**Review by mbrobbel — COMMENTED** [review]
Thanks @dhruv9vats!

For your PR description: to get it to auto-close the referenced issue you should use one of [these keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

File: src/offset.rs (hunk 1 above)
Comment (dhruv9vats) [reply src/offset.rs]: It looks much cleaner with `Peekable`, thanks for the suggestion!

File: src/offset.rs [hunk 2]
```diff
@@ -490,6 +490,125 @@ where
...
+}
+
+impl<T, OffsetItem: OffsetElement, Buffer: BufferType> IntoIterator
+    for Offset<T, true, OffsetItem, Buffer>
+where
+    T: IntoIterator,
+    <Buffer as BufferType>::Buffer<OffsetItem>: Validity<
+        true,
+        Storage<Buffer> = Nullable<<Buffer as BufferType>::Buffer<OffsetItem>, Buffer>,
+    >,
```
[the remark this answers is not in the document]
Comment (dhruv9vats) [reply src/offset.rs]: Yeah, I might have typed it out before `clippy` finished its job :P

File: src/offset.rs [hunk 3]
```diff
@@ -490,6 +491,108 @@ where
...
+impl<T, OffsetItem: OffsetElement, Buffer: BufferType> Iterator
+    for OffsetIntoIter<T, OffsetItem, Buffer>
+where
+    T: IntoIterator,
+    <Buffer as BufferType>::Buffer<OffsetItem>: IntoIterator<Item = OffsetItem>,
+{
+    type Item = Vec<<T as IntoIterator>::Item>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let (Some(start), Some(&end)) = (self.offsets.next(), self.offsets.peek()) {
```
Comment (mbrobbel) [inline src/offset.rs:515]: Nit: I think it's a bit more clear if you use:
```rust
self.offsets
  .next()
  .and_then(|start| self.offsets.peek().map(|end| (start, *end)))
  .map(|(start, end)| { ... })
```

File: src/offset.rs [hunk 4]
```diff
@@ -490,6 +491,108 @@ where
...
+            let slot_end: usize = end.try_into().expect("offset value should be in range");
+
+            assert!(
+                slot_end >= slot_start,
+                "offsets must be monotonically increasing, but a slot's start \
+                offset value {slot_start} is greater than the slot's end offset \
+                value {slot_end}, resulting in a negative slot length"
+            );
+
+            let slot_length = slot_end - slot_start;
```
Comment (mbrobbel) [inline src/offset.rs:526]: Nit: maybe use https://doc.rust-lang.org/stable/std/primitive.usize.html#method.checked_sub?

File: src/offset.rs [hunk 5]
```diff
@@ -490,6 +491,108 @@ where
...
+            let taken = self.data.by_ref().take(slot_length).collect::<Vec<_>>();
+
+            assert!(
```
Comment (mbrobbel) [inline src/offset.rs:529]: We can assume correct inputs in these impls:
```suggestion
            debug_assert!(
```

File: src/offset.rs [hunk 6]
```diff
@@ -490,6 +491,108 @@ where
...
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.offsets
+            .validity
+            .into_iter()
+            // We use zip insead of map because a null value may have a
+            // positive slot length. That is, a null value may occupy a
+            // non-empty memory space in the data buffer.
+            // Thus the underlying iterator must be advanced even for null
+            // values as indicated by the validity bitmap.
```
Comment (mbrobbel) [inline src/offset.rs:587]: Can you add a test for this? I think the easiest way to build an example array is to use `FromIterator` and then flip some validity bits.

Comment (dhruv9vats) [reply src/offset.rs:583]: added one in [fix: use iter map and add test for non-empty nulls](https://github.com/mbrobbel/narrow/pull/186/commits/29a92cf2dbd1e88a57ec30a9db7e56c53cdde02f)

File: src/offset.rs [hunk 7]
```diff
@@ -490,6 +491,104 @@ where
...
+            .zip(OffsetIntoIter {
+                data: self.data.into_iter(),
+                offsets: self.offsets.data.into_iter().peekable(),
+            })
+            .map(|(validity, value): (bool, Vec<_>)| validity.then_some(value))
```
Comment (mbrobbel) [inline src/offset.rs:588]: Maybe you should add a comment that we need the eager evaluation of `then_some` here?

File: src/offset.rs [hunk 8]
```diff
@@ -625,25 +724,53 @@ mod tests {
...
+            offset_nullable.offsets.bitmap_ref_mut().is_valid(0),
+            Some(false)
+        );
+        // while there's non-empty data in the underlying data array
+        assert_eq!(offset_nullable.data.as_slice(), &[1, 2, 3, 4, 5, 6, 7, 8]);
+
+        assert_eq!(
+            offset_nullable.into_iter().collect::<Vec<_>>(),
+            // the first item in not yielded even though there's data for it
+            // because the validity bitmat was altered.
```
Comment (mbrobbel) [inline src/offset.rs:772]:
```suggestion
            // because the validity bitmap was altered.
```

**Review by mbrobbel — COMMENTED** [review]
Final nits.

File: src/offset.rs (hunk 7 above)
Comment (dhruv9vats) [reply src/offset.rs:588]: Could you elaborate a bit? I'm not sure if I fully understand? Since the underlying iterator must always be advanced.

Comment (mbrobbel) [reply src/offset.rs:588]: Oh right, the `data` iterator is advanced by `OffsetIntoIter`. We should maybe consider special-casing this to avoid the allocation of the `Vec` for `null` values using something like https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.advance_by when "skipping" over a null element. But that's an optimization we can do in a follow-up PR.

**Review by mbrobbel — APPROVED** [review]
